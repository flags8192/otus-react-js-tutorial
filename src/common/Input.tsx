import React, { PureComponent } from 'react';
import styled from '@emotion/styled';

const Root = styled.div`
  display: flex;
  input {
    margin-left: 10px;
    width: 500px;
  }
`;

interface IInputProps {
  onChange: (val: string) => void;
  autoFocus?: boolean;
}
interface IInputState {
  value: string;
}
export default class Input extends PureComponent<IInputProps, IInputState> {
  input: React.RefObject<HTMLInputElement>;

  constructor(props: IInputProps) {
    super(props);

    this.state = {
      value: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.input = React.createRef();
  }

  componentDidMount() {
    window.addEventListener('keydown', this.handleKeydown);
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.handleKeydown);
  }

  componentDidUpdate(prevProps: {}, prevState: IInputState) {
    const oldValue = prevState.value.trimStart();
    const newValue = this.state.value.trimStart();
    if (oldValue !== newValue) {
      this.props.onChange(newValue);
    }
  }

  handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ value: e.target.value });
  }

  handleKeydown = () => {
    if (
      this.props.autoFocus &&
      this.input.current &&
      document.activeElement !== this.input.current
    ) {
      this.input.current.focus();
    }
  };

  render() {
    return (
      <Root>
        Поиск:
        <input
          role="filterInput"
          ref={this.input}
          autoFocus={this.props.autoFocus}
          onChange={this.handleChange}
        />
      </Root>
    );
  }
}
