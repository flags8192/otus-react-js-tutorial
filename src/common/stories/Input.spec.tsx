import React from 'react';
import { render, screen } from '@testing-library/react';
import Input from '@src/common/Input';

describe('Input', () => {
  it('Render <Input>', () => {
    const handleChange = jest.fn();
    const { asFragment } = render(<Input onChange={handleChange} autoFocus={true} />);
    expect(asFragment()).toMatchSnapshot();
    const filterInput = screen.getByRole(/filterInput/gi);
    expect(filterInput).toBeInTheDocument();
  });
});
