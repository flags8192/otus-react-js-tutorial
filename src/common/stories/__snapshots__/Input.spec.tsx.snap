// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Render <Input> 1`] = `
<DocumentFragment>
  <div
    class="css-1ng1nrc"
  >
    Поиск:
    <input
      role="filterInput"
    />
  </div>
</DocumentFragment>
`;
